p384/src/arithmetic/hash2curve.rs

```diff
@@ -0,0 +1,148 @@
+use elliptic_curve::bigint::{ArrayEncoding, U384};
+use elliptic_curve::consts::U72;
+use elliptic_curve::generic_array::GenericArray;
+use elliptic_curve::group::cofactor::CofactorGroup;
+use elliptic_curve::hash2curve::{FromOkm, GroupDigest, MapToCurve};
+use elliptic_curve::subtle::{Choice, CtOption};
+
+use crate::{NistP384, ProjectivePoint, Scalar};
+
+use super::FieldElement;
+
+impl GroupDigest for NistP384 {
+    type FieldElement = FieldElement;
+}
+
+impl FromOkm for FieldElement {
+    type Length = U72;
+
+    fn from_okm(_data: &GenericArray<u8, Self::Length>) -> Self {
+        todo!()
+    }
+}
+
+impl MapToCurve for FieldElement {
+    type Output = ProjectivePoint;
+
+    fn map_to_curve(&self) -> Self::Output {
+        todo!()
+    }
+}
+
+impl CofactorGroup for ProjectivePoint {
+    type Subgroup = ProjectivePoint;
+
+    fn clear_cofactor(&self) -> Self::Subgroup {
+        *self
+    }
+
+    fn into_subgroup(self) -> CtOption<Self::Subgroup> {
+        CtOption::new(self, 1.into())
+    }
+
+    fn is_torsion_free(&self) -> Choice {
+        1.into()
+    }
+}
+
+impl FromOkm for Scalar {
+    type Length = U72;
+
+    fn from_okm(data: &GenericArray<u8, Self::Length>) -> Self {
+        const F_2_288: Scalar = Scalar(U384::from_be_hex(
+            "000000000000000000000001000000000000000000000000000000000000000000000000000000000000000000000000",
+        ));
+
+        let mut d0 = GenericArray::default();
+        d0[12..].copy_from_slice(&data[0..36]);
+        let d0 = Scalar(U384::from_be_byte_array(d0));
+
+        let mut d1 = GenericArray::default();
+        d1[12..].copy_from_slice(&data[36..]);
+        let d1 = Scalar(U384::from_be_byte_array(d1));
+
+        d0 * F_2_288 + d1
+    }
+}
+
+#[test]
+fn hash_to_scalar_voprf() {
+    use elliptic_curve::hash2curve::ExpandMsgXmd;
+    use hex_literal::hex;
+    use sha2::Sha384;
+
+    struct TestVector {
+        dst: &'static [u8],
+        seed: &'static [u8],
+        sk_sm: &'static [u8],
+    }
+
+    const TEST_VECTORS: &[TestVector] = &[
+        TestVector {
+            dst: b"HashToScalar-VOPRF08-\x00\x00\x04",
+            seed: &hex!("a3a3a3a3a3a3a3a3a3a3a3a3a3a3a3a3a3a3a3a3a3a3a3a3a3a3a3a3a3a3a3a3a3a3a3a3a3a3a3a3a3a3a3a3a3a3a3a3"),
+            sk_sm: &hex!("b9ff42e68ef6f8eaa3b4d15d15ceb6f3f36b9dc332a3473d64840fc7b44626c6e70336bdecbe01d9c512b7e7d7e6af21"),
+        },
+        TestVector {
+            dst: b"HashToScalar-VOPRF08-\x01\x00\x04",
+            seed: &hex!("a3a3a3a3a3a3a3a3a3a3a3a3a3a3a3a3a3a3a3a3a3a3a3a3a3a3a3a3a3a3a3a3a3a3a3a3a3a3a3a3a3a3a3a3a3a3a3a3"),
+            sk_sm: &hex!("42c4d1c15d27be015844404088967afe48c8ae96d4f00ce48e4d38ecabfb8feb5b748de625cdf81ab076745d6211be95"),
+        },
+    ];
```

Comment:
I'm using test Voprf version 8 test vectors to keep it simple, as version 9 requires more code.

Reply:
Now that everything's working, I updated the test vectors.